partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/Operator.kt

```diff
@@ -14,17 +15,15 @@ internal sealed interface Operator {
     interface Expr : Operator {
 
         @OptIn(PartiQLValueExperimental::class)
-        fun eval(record: Record): PartiQLValue
+        fun eval(env: Environment): PartiQLValue
     }
 
     /**
      * Relation operator represents an evaluable collection of binding tuples.
      */
-    interface Relation : Operator, AutoCloseable {
+    interface Relation : Operator, AutoCloseable, Iterator<Record> {
 
-        fun open()
-
-        fun next(): Record?
+        fun open(env: Environment)
```

Comment:
No. Even if we defer materialization until `next` or `hasNext`, we'll need to pass an environment somehow.

Take for instance PartiQL's lateral join. The join operator will loop over all elements of the LHS and evaluate them in the same environment as the JOIN. However, for each element on the LHS, the join operator will open the RHS, loop through its elements, and close the RHS -- all while giving the RHS a new environment. That new environment would be the original concatenated with the record received from the LHS.

There currently isn't a mechanism to close over the new environment. `Iterator`'s `hasNext` and `next` wouldn't allow us to pass a new environment to the RHS of a JOIN. The previous implementation didn't need to handle the closing over environments because it was immediately materialized, and lazy materialization has issues with the global environment.

Reply:
Oh no I see now. Open has a different meaning now compared to volcano because you can't pass the environment in `hasNext()/next()` you have to make the environment available. But the compiler is set up to do all instantiation all operators once.

With lazy expressions we have to close over the environment or else it will go out of scope, hence the creation of the "thunks" which are really just closures of the bindings environment that return operator implementations. This is very slow because you have to instantiate a new operator (for everything) on each new record.

 Here we are instantiating everything once in the compiler, then updating the reference to the environment via open (if I understand correctly now).

Reply:
Yes, exactly